### PR TITLE
cloud-storage-contrib-nio: Add support for newFileChannel

### LINKED ISSUE #4702: google-cloud-nio: CloudStorageFileSystemProvider throws UnsupportedException when used with Webflux + MultipartFile upload
Because `CloudStorageFileSystemProvider` does not override the `newFileChannel` method it throws an `UnsupportedOperationException` when used for Multipart file uploads with `FilePart.transferto(Path path)` in a reactive Spring boot application.

#### Environment details

1. Specify the API: google-cloud-nio
2. OS type and version: Debian 9
3. Java version: 10
4. google-cloud-nio version(s): 0.83-alpha

#### Steps to reproduce

  1. See code sample

#### Code example

```java
    @PostMapping(value = "/upload", consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
    public Mono<Void> multipartUpload( final @RequestPart("file") FilePart filePart
    ) {

        FileSystem fileSystem = CloudStorageFileSystem.forBucket("my-bucket");
        return filePart.transferTo(fileSystem.getPath("failed-file"));
    }
```

#### Stack trace
```
java.lang.UnsupportedOperationException: null
    at java.base/java.nio.file.spi.FileSystemProvider.newFileChannel(FileSystemProvider.java:524)
    at java.base/java.nio.channels.FileChannel.open(FileChannel.java:292)
    at java.base/java.nio.channels.FileChannel.open(FileChannel.java:345)
    at org.springframework.http.codec.multipart.SynchronossPartHttpMessageReader$SynchronossFilePart.transferTo(SynchronossPartHttpMessageReader.java:334)
```

#### External references such as API reference guides used	

- Stackoverflow question - https://stackoverflow.com/questions/55246844/unsupportedoperationexception-with-multipart-file-upload-using-webflux-and-sprin/55248459#55248459
- Java documentation for FileServiceProvider - https://docs.oracle.com/javase/7/docs/api/java/nio/file/spi/FileSystemProvider.html#newFileChannel(java.nio.file.Path,%20java.util.Set,%20java.nio.file.attribute.FileAttribute...)	
- FilePart javadoc - https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/http/codec/multipart/FilePart.html#transferTo-java.nio.file.Path-

### PR DESCRIPTION
**Pull request by olavloite**
Adds support for the method FileSystemProvider.newFileChannel(Path, Set<? extends OpenOption>,      FileAttribute<?>...) to Google Cloud Storage nio. This enables, among others, users to use google-cloud-storage-nio with Webflux and MultipartFile upload.

Fixes #4702

### CODE REVIEW

**Comment by sduskis** [comment]
Can you please run this: `mvn com.coveo:fmt-maven-plugin:format`

File: google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageReadFileChannel.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright 2016 Google LLC
```
Comment (kolea2) [inline google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageReadFileChannel.java]: Please update the year.

Comment (olavloite) [reply google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageReadFileChannel.java]: changed

File: google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemProvider.java [hunk 2]
```diff
@@ -300,6 +301,51 @@ public SeekableByteChannel newByteChannel(
...
+   *   <li>{@link StandardOpenOption#WRITE}
+   *   <li>{@link StandardOpenOption#TRUNCATE_EXISTING}
+   * </ul>
+   *
+   * In all other cases the {@link FileChannel} will be read-only.
+   *
+   * @param path The path to the file to open or create
+   * @param options The options specifying how the file should be opened, and whether the {@link
+   *     FileChannel} should be read-only or write-only.
+   * @param attrs (not supported, the values will be ignored)
```
Comment (JesseLovelace) [inline google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemProvider.java]: Why include attrs here at all if this is the case? attrs is optional in File.createFile

Comment (olavloite) [reply google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemProvider.java]: I chose to do so to keep it consistent with the other methods in `CloudStorageFileSystemProvider` that accept an array of `FileAtribute`s as an input parameter. These do the exact same check, even though the input value is ignored.

(See `newByteChannel(...)``createDirectory(...)`)

Furthermore, `newFileChannel` method can also be used to read an existing file, so it does not always call File.createFile.

File: google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemProvider.java [hunk 3]
```diff
@@ -300,6 +301,51 @@ public SeekableByteChannel newByteChannel(
...
+    }
+    if (options.contains(StandardOpenOption.WRITE)
+        || options.contains(StandardOpenOption.CREATE)
+        || options.contains(StandardOpenOption.CREATE_NEW)
+        || options.contains(StandardOpenOption.TRUNCATE_EXISTING)) {
+      CloudStorageWriteChannel writeChannel =
+          (CloudStorageWriteChannel) newWriteChannel(path, options);
+      return new CloudStorageWriteFileChannel(writeChannel);
+    } else {
+      CloudStorageReadChannel readChannel = (CloudStorageReadChannel) newReadChannel(path, options);
```
[the remark this answers is not in the document]
Comment (olavloite) [reply google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemProvider.java]: Good point, changed.

File: google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageReadFileChannel.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+  @Override
+  public long transferFrom(ReadableByteChannel src, long position, long count) throws IOException {
+    throw new UnsupportedOperationException(READ_ONLY);
+  }
+
+  @Override
+  public int read(ByteBuffer dst, long position) throws IOException {
+    synchronized (this) {
+      long originalPosition = position();
+      position(position);
```
Comment (sduskis) [inline google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageReadFileChannel.java]: Isn't this redundant?

```
      long originalPosition = position();
      position(position);
```

Do we actually need `position(position);`?

File: google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageWriteFileChannel.java [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+      checkValid();
+      return writeChannel.position();
+    }
+  }
+
+  @Override
+  public FileChannel position(long newPosition) throws IOException {
+    synchronized (this) {
+      checkValid();
+      if (newPosition != position()) {
```
Comment (sduskis) [inline google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageWriteFileChannel.java]: There's a call to `checkValid()` in `position()`.  We don't need two calls.  Perhaps just use `writeChannel.position()`

Comment (sduskis) [reply google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageWriteFileChannel.java]: There are other places where the double check happens as well, further down.

File: google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageReadFileChannel.java (hunk 4 above)
Comment (olavloite) [reply google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageReadFileChannel.java]: Yes, we need to position the channel where the read should begin. I've changed the name of the parameter to make it clearer.

File: google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageReadFileChannel.java [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+  @Override
+  public int read(ByteBuffer dst) throws IOException {
+    return readChannel.read(dst);
+  }
+
+  @Override
+  public long read(ByteBuffer[] dsts, int offset, int length) throws IOException {
+    long res = 0L;
+    synchronized (this) {
```
[the remark this answers is not in the document]
Comment (olavloite) [reply google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageReadFileChannel.java]: Changed.

File: google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageReadFileChannel.java [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+
+  @Override
+  public void force(boolean metaData) throws IOException {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public long transferTo(long position, long count, WritableByteChannel target) throws IOException {
+    long res = 0L;
+    synchronized (this) {
```
[the remark this answers is not in the document]
Comment (olavloite) [reply google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageReadFileChannel.java]: Changed.

File: google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageReadFileChannel.java [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+  }
+
+  @Override
+  public long transferFrom(ReadableByteChannel src, long position, long count) throws IOException {
+    throw new UnsupportedOperationException(READ_ONLY);
+  }
+
+  @Override
+  public int read(ByteBuffer dst, long position) throws IOException {
+    synchronized (this) {
```
[the remark this answers is not in the document]
Comment (olavloite) [reply google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageReadFileChannel.java]: Changed.

File: google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageWriteFileChannel.java (hunk 5 above)
Comment (olavloite) [reply google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageWriteFileChannel.java]: I removed the redundant validation checks and changed the method signatures in this class as well to synchronized.

**Pull request merged**